tests/test_transactions.py

```diff
@@ -333,13 +358,131 @@ async def test_task():
+                PREFECT_TASK_SCHEDULING_DEFAULT_STORAGE_BLOCK: f"local-file-system/{name}",
+            }
+        ):
+            yield
+
+    @pytest.fixture
+    async def result_1(self, default_storage_setting):
+        result_factory = await ResultFactory.default_factory(persist_result=True)
+        return await result_factory.create_result(obj={"foo": "bar"})
+
+    @pytest.fixture
+    async def result_2(self, default_storage_setting):
+        result_factory = await ResultFactory.default_factory(persist_result=True)
+        return await result_factory.create_result(obj={"fizz": "buzz"})
+
+    async def test_basic_transaction(self, result_1):
+        store = MemoryRecordStore()
+        with transaction(key="test_basic_transaction", store=store) as txn:
+            assert isinstance(txn.store, MemoryRecordStore)
+            txn.stage(result_1)
+
+        result_1 = txn.read()
+        assert result_1
+        assert await result_1.get() == {"foo": "bar"}
+
+        record = store.read("test_basic_transaction")
+        assert record
+        assert record.result == result_1
+        assert record.key == "test_basic_transaction"
+
+    async def test_competing_read_transaction(self, result_1):
+        transaction_1_open = threading.Event()
+        transaction_2_open = threading.Event()
+
+        def competing_transaction(
+            store, key, result, transaction_1_open, transaction_2_open
+        ):
+            # isolation level is SERIALIZABLE, so a lock will be taken
+            with transaction(
+                key=key, store=store, isolation_level=IsolationLevel.SERIALIZABLE
+            ) as txn:
+                transaction_1_open.set()
+                transaction_2_open.wait()
+                # artificial delay to ensure the other transaction waits on read
+                sleep(1)
+                txn.stage(result)
+
+        store = MemoryRecordStore()
+        thread = threading.Thread(
+            target=competing_transaction,
+            args=(
+                store,
+                "test_competing_read_transaction",
+                result_1,
+                transaction_1_open,
+                transaction_2_open,
+            ),
+        )
+        thread.start()
+        transaction_1_open.wait()
+        with transaction(key="test_competing_read_transaction", store=store) as txn:
+            transaction_2_open.set()
+            read_result = txn.read()
+
+        assert read_result == result_1
+        thread.join()
+
+    async def test_competing_write_transaction(self, result_1, result_2):
+        transaction_1_open = threading.Event()
+
+        def competing_transaction(store, key, result, transaction_1_open):
+            with transaction(
+                key=key, store=store, isolation_level=IsolationLevel.SERIALIZABLE
+            ) as txn:
+                transaction_1_open.set()
+                # artificial delay to ensure the other transaction waits on open
+                sleep(1)
```

Comment:
If this test ever flakes, you can replace this `sleep` with `wait`ing on another event that the outer function `set`s.  That would make this test deterministic and also run faster than 1s

Reply:
I think this is left over from when I wanted to ensure this was a failing test first (sometimes it would flake in the other direction and pass), but I don't think we need it long-term. I removed it in  https://github.com/PrefectHQ/prefect/pull/14939/commits/d196cfe02c320cf59b5672053b8a81b28cba9095.